mmdet/core/hook/neptunelogger_hook.py

```diff
@@ -0,0 +1,331 @@
+        for key, value in runner.log_buffer.val_history.items():
+            self.base_handler['train/' + category + '/' + key].append(
+                value[-1])
+
+    def _log_checkpoint(self,
+                        runner,
+                        ext='pth',
+                        final=False,
+                        mode='epoch') -> None:
+        assert mode in ['epoch', 'iter']
+
+        if self.ckpt_hook is None:
+            return
+
+        if final:
+            file_name = 'latest'
+        else:
+            file_name = f'{mode}_{getattr(runner, mode) + 1}'
+
+        file_path = osp.join(self.ckpt_hook.out_dir, f'{file_name}.{ext}')
+
+        neptune_checkpoint_path = osp.join('model/checkpoint/',
+                                           f'{file_name}.{ext}')
+
+        if not osp.exists(file_path):
+            runner.logger.warning(
+                f'Checkpoint {file_path} not found - skipping.')
+            return
+        with open(file_path, 'rb') as fp:
+            self._run[neptune_checkpoint_path] = File.from_stream(fp)
+
+    @master_only
+    def after_train_iter(self, runner) -> None:
+        """For an iter-based runner logs evaluation metadata, as well as
+        checkpoints (if enabled by the user)."""
+        if not isinstance(runner, IterBasedRunner):
+            return
+
+        log_eval = self.every_n_iters(runner, self.eval_hook.interval)
+        if log_eval:
+            self._log_evaluation(runner, 'iter')
+
+        self._log_buffer(runner, 'iter')
+        if self._should_upload_checkpoint(runner):
+            self._log_checkpoint(runner, final=False, mode='iter')
+
+    def _should_upload_checkpoint(self, runner) -> bool:
+        if isinstance(runner, EpochBasedRunner):
+            return self.log_checkpoint and \
+                (runner.epoch + 1) % self.ckpt_interval == 0
+        elif isinstance(runner, IterBasedRunner):
+            return self.log_checkpoint and \
+                (runner.iter + 1) % self.ckpt_interval == 0
+        else:
+            return False
+
+    @master_only
+    def after_train_epoch(self, runner) -> None:
+        """For an epoch-based runner logs evaluation metadata, as well as
+        checkpoints (if enabled by the user)."""
+        if not isinstance(runner, EpochBasedRunner):
+            return
+        log_eval = self.every_n_epochs(runner, self.eval_hook.interval)
+        if log_eval:
+            self._log_evaluation(runner, 'epoch')
+
+        self._log_buffer(runner, 'epoch')
+
+        if self._should_upload_checkpoint(runner):
+            self._log_checkpoint(runner, final=False)
+
+    @master_only
+    def after_run(self, runner) -> None:
+        """If enabled by the user, logs final model checkpoint, syncs and stops
+        the Neptune run."""
+        if self.log_model:
+            self._log_checkpoint(runner, final=True)
+
+        runner.logger.info('Syncing with Neptune.ai')
+        self._run.sync()
+        self._run.stop()
```

Comment:
Does stop automatically call `sync`? In that case, we won't need the `sync` before.